### PR TITLE
Za/806 analyst view domain management data

### LINKED ISSUE #806: Analyst view of domain management data
### Story

As an analyst
I want to be able to view all data entered in the domain management pages for a given domain
so that I can easily see and modify contact, host and other dns info 

### Acceptance Criteria

- For approved domains, analysts can view all domain management information clearly (see the domain overview page example in additional context)
- This data should be editable
- Optional: have some data editable while others are not (requires refinement and can be added on later).

If option 2 (see below) is implemented:
- The button in Django admin should read “Manage Domain”
- The domain management page should open in a new tab. Within that tab, remove the “Back to manage your domains” button (for analysts only) 
- Show an alert message at the top of the Domain Management page telling analyst that they are viewing someone else’s domain and should close the tab when finished. Gaby created a [Figma mock-up to show what this could look like](https://www.figma.com/file/aJbuDLJNNikqQObTuCNKQa/GetGov_Designs_ECS-Truss?type=design&node-id=491%3A26602&mode=design&t=uTSNRvle2VbI9pA4-1).

### Additional Context

It is important to be able to see which contacts are associated with a given domain's management, especially when needing to reach out to them. In the current system, analysts can view all domain management data but can only edit some data.

- One option is to make a view on django admin to show all this data nicely to match the domain overview page. Most data is already able to be retrieved from Domain Information, the exception is the epp centric stuff.
- Option 2 is to have a button on djano admin domain page link the analyst to the correct domain management page. Here it would be good to add in logging to indicate that the actions are being carried out by an analysts. 

for context, this is the domain management page that users see and which has all the data that an analyst needs to see. 
image

for refinement:
1. would option 2 be confusing for analysts as the domain management page has a section called "Your Contact information"
2.) is this going to require using the Public Contact class?

### Issue Links
relates to: #807 
_No response_

### PR DESCRIPTION
**Pull request by zandercymatics**
## Ticket

Resolves #806

## Changes

- A button on django admins domain page, labeled 'Manage Domains'. When this button is clicked, it redirects to the domain management page (which opens in a new tab).
- If the user has adequate permissions, if they clicked this button, and if its the same domain, then the page will display an alert
- ~Logging actions for superusers and analysts who do this in console~ This was removed as it was determined to be redundant by Rachid (as we are already logging in /admin). Basically, we are already logging these actions within that context
- Changes to html pages: domain_change_form.html, domain_base.html
- Permission tweaks to facilitate this change
- Adds two session variables, "analyst_action" and "analyst_action_location" which keep track of this
- Added get-gov-admin.js as I could not make a template change without breaking the 'placehold' button when opening in a new tab. Contains a small script that facilitates that. Basically a template for future use. Let me know if it is better to have this exist in get-gov.js instead, but I figured that we would not want to share that same script in future if we need to add some js to the admin area specifically.
- Added test cases

## Context for reviewers

- Given this particular execution, is this something we should test further, and if so, what should fall under that scope?
- What fields should an analyst not be able to edit? Not able to view?
- Should text fields have their verbiage changed to reflect that the analyst is not the original creator?
- Any issues with this specific implementation? 
- For the alert, the Figma has some margin differences between the domain_sidebar and the domain_detail page (the navigation bar on the left and the content on the right). This causes small differences in how much text content can be displayed. In addition to that, the default styling for the 'warning' icon is positioned differently as opposed to the figma. 
- For those two figma differences, should those be fixed? Both changes will make minor styling differences app-wide if we want to keep consistency 
- What should be the level of granularity when it comes to logging? Is how I have it okay, or does it need to show a deeper breakdown?

## Setup
Navigate to https://getgov-za.app.cloud.gov/ and login, then as both an analyst and admin, navigate to /admin, and click 'domains'. You should see a 'Manage Domain' button near 'place hold', click that to view this behavior.

#### Satisfied acceptance criteria and met development standards

- [x] Met the acceptance criteria, or will meet them in a subsequent PR
- [x] Created/modified automated tests
- [x] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [ ] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [x] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [ ] Interactions with external systems are wrapped in try/except
- [x] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [x] Checked keyboard navigability
- [x] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [x] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [x] Pulled this branch locally and tested it
- [x] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
- [x] Made it clear which comments need to be addressed before this work is merged
- [x] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [x] Interactions with external systems are wrapped in try/except
- [x] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [x] Checked that the design translated visually
- [x] Checked behavior
- [x] Checked different states (empty, one, some, error)
- [x] Checked for landmarks, page heading structure, and links
- [x] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [x] Checked keyboard navigability
- [x] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [x] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

## Screenshots

image
image

### CODE REVIEW

**Comment by rachidatecs** [comment]
On tests: 
1- Blocking: test_session_variables_retain_information is failing
2- For better coverage (and more robust testing in general), it would be nice to set up the right (and wrong) conditions, trigger the redirect and test the response. This is a strong nice-to-have imo

**Comment by zandercymatics** [comment]
@gabydisarli @Katherine-Osos  This should be good for you guys to test

**Comment by Katherine-Osos** [comment]
@zandercymatics This looks good to me; however, @gabydisarli is reviewing as well so she may have comments before you merge.

One note, there's a period missing at the end of the alert message on the Domain Overview.

**Review by Katherine-Osos — APPROVED** [review]

**Comment by zandercymatics** [comment]
@rachidatecs @therealslimhsiehdy  Everything look good on your guys' end?

**Review by gabydisarli — APPROVED** [review]
I wasn't able to test as a analyst user since I'm waiting on my PR to pass, but I've checked off all other design review tasks I could. Looks good to me!

File: src/registrar/fixtures.py [hunk 1]
```diff
@@ -63,7 +63,7 @@ class UserFixture:
             "last_name": "Adkinson",
         },
         {
-            "username": "bb21f687-c773-4df3-9243-111cfd4c0be4",
+            "username": "2bf518c2-485a-4c42-ab1a-f5a8b0a08484",
```
[the remark this answers is not in the document]
Comment (zandercymatics) [reply src/registrar/fixtures.py:66]: Oof I did not mention that... Yes Paul asked me to change his uuid so he can see stuff on my sandbox. Guess it snuck in here.... When I merge this I'll undo that change and separate it out into a separate PR, nice catch!

File: src/registrar/views/utility/mixins.py [hunk 2]
```diff
@@ -25,25 +33,81 @@ def has_permission(self):
...
+        # If pk is none then something went very wrong...
+        if pk is None:
+            raise ValueError("Primary key is None")
+
+        # ticket 806
+        if self.can_access_other_user_domains(pk):
+            return True
+
+        # user needs to have a role on the domain,
+        # and user cannot be restricted
```
Comment (rachidatecs) [inline src/registrar/views/utility/mixins.py:52]:
```suggestion
```

File: src/registrar/views/utility/mixins.py [hunk 3]
```diff
@@ -25,25 +33,81 @@ def has_permission(self):
...
+    def can_access_other_user_domains(self, pk):
+        """Checks to see if an authorized user (staff or superuser)
+        can access a domain that they did not create or was invited to.
+        """
+
+        # Check if the user is permissioned...
+        user_is_analyst_or_superuser = (
+            self.request.user.is_staff or self.request.user.is_superuser
+        )
+        logger.debug(f"is auth {user_is_analyst_or_superuser}")
```
Comment (rachidatecs) [inline src/registrar/views/utility/mixins.py:70]: Do we want to keep this log?

File: src/registrar/views/utility/mixins.py [hunk 4]
```diff
@@ -25,25 +33,81 @@ def has_permission(self):
...
+        # In other words, if the analyst/admin did not click
+        # the 'Manage Domain' button in /admin,
+        # then they cannot access this page.
+        session = self.request.session
+        can_do_action = (
+            "analyst_action" in session
+            and "analyst_action_location" in session
+            and session["analyst_action_location"] == pk
+        )
+        logger.debug(f"can do {can_do_action}")
```
Comment (rachidatecs) [inline src/registrar/views/utility/mixins.py:85]: Same

File: src/registrar/views/utility/permission_views.py [hunk 5]
```diff
@@ -27,6 +31,77 @@ class DomainPermissionView(DomainPermission, DetailView, abc.ABC):
...
+        action_location = "analyst_action_location"
+        # Flag to see if an analyst is attempting to make edits
+        if action in self.request.session:
+            context[action] = self.request.session[action]
+        if action_location in self.request.session:
+            context[action_location] = self.request.session[action_location]
+
+        return context
+
+    def log_analyst_form_actions(
```
Comment (rachidatecs) [inline src/registrar/views/utility/permission_views.py:50]: I feel like this does not belong in DomainPermissionView, but rather in some DomainLoggingClass. Furthermore, we're using Audit Log which looks like it's already logging this info.

I would gut this out and move it into another PR, and park it for further investigation.

File: src/registrar/templates/django/admin/domain_change_form.html [hunk 6]
```diff
@@ -1,7 +1,14 @@
 {% extends 'admin/change_form.html' %}
+{% load i18n static %}
+
+{% block extrahead %}
+{{ block.super }}
+<script type="text/javascript" src="{% static 'js/get-gov-admin.js' %}" defer></script>
```
[the remark this answers is not in the document]
Comment (zandercymatics) [reply src/registrar/templates/django/admin/domain_change_form.html]: This has to do with the structure of the template... Basically this js file is what allows it to open in another tab. There is not really a method in Django to redirect to another tab as I believe that is handled by the browser.

All of the 'buttons' on this page (mostly placehold) are nested in a shared form object. Without it, I'd basically need to do a somewhat extensive form + django rewrite which I'd like to avoid doing if possible as I do not know if that will break things. That said - I think we may be able to avoid this js implementation (potentially in another PR given that it looks like we're adding more buttons to this page in general), what are you thoughts on that?

File: src/registrar/views/utility/mixins.py [hunk 7]
```diff
@@ -25,26 +33,70 @@ def has_permission(self):
...
-            user=self.request.user, domain__id=self.kwargs["pk"]
+        if UserDomainRole.objects.filter(
+            user=self.request.user, domain__id=pk
         ).exists():
+            return True
+
+        # ticket 806
+        requested_domain: DomainInformation = None
+        try:
+            requested_domain = DomainInformation.objects.get(id=pk)
```
[the remark this answers is not in the document]
Comment (zandercymatics) [reply src/registrar/views/utility/mixins.py]: Not sure if this is the right variable to begin with, but Domain.statuses isn't implemented yet. That said, I think I can grab a domain_application object directly - if I recall I think I moved it up a layer due to pk differences but I should retest this to see if this is applicable still (leftover code, essentially)

File: src/registrar/assets/js/get-gov-admin.js
Comment (therealslimhsiehdy) [inline src/registrar/assets/js/get-gov-admin.js:1]: NIT/non-blocking: I had the [same question](https://github.com/cisagov/getgov/pull/928#discussion_r1306149694) as Rachid when looking at this PR 😅

Might be worth adding a comment mentioning "bc of structure of [template name here], this js file allows us to redirect the changes into a new tab bc of [reason here]" for future reference! YMMV

**Review by therealslimhsiehdy — APPROVED** [review]
LGTM -- strong work on the refactor + testing of the sessions + comments + everything really!
